File: from-0401-to-0500/0463-island-perimeter/solution.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        
        visited = set()
        
        def dfs(row, col):
            nonlocal grid, m, n, visited
            if row < 0 or col < 0 or row >= m or col >= n or grid[row][col] == 0:
                return 1
            if (row, col) in visited:
                return 0
            visited.add((row, col))

            perim = dfs(row + 1, col)
            perim += dfs(row - 1, col)
            perim += dfs(row, col + 1)
            perim += dfs(row, col - 1)
            
            return perim
        
        for row in range(m):
            for col in range(n):
                if grid[row][col] == 1:
                    return dfs(row, col)
                
        return 0
            

File: from-0401-to-0500/0463-island-perimeter/solution.go
type Cell struct {
    row int
    col int
}

func islandPerimeter(grid [][]int) int {
    m, n := len(grid), len(grid[0])
    seen := make(map[Cell]bool)

    var dfs func(row, col int) int
    dfs = func(row, col int) int {
        if row < 0 || row >= m || col < 0 || col >= n || grid[row][col] == 0 {
            return 1
        }
        cell := Cell{row: row, col: col}
        if seen[cell] {
            return 0
        }
        seen[cell] = true
        perimeter := dfs(row + 1, col)
        perimeter += dfs(row - 1, col)
        perimeter += dfs(row, col + 1)
        perimeter += dfs(row, col - 1)
        return perimeter
    }
    for r := 0; r < m; r ++ {
        for c := 0; c < n; c ++ {
            if grid[r][c] == 1 {
                return dfs(r, c)
            }
        }
    }
    return 0
}